File: coding-test/programmers/src/kr/co/programmers/kakao_blind_recruitment_2021/NewIdRecommendation.java
package kr.co.programmers.kakao_blind_recruitment_2021;

public class NewIdRecommendation {

    public static void main(String[] args) {
        String new_id = "...!@BaT#*..y.abcdefghijklm";
        System.out.println(new Solution().solution(new_id));
    }

    static class Solution {
        public String solution(String new_id) {
            String answer = "";

            answer = new_id.toLowerCase();

            answer = answer.replaceAll("[^a-z0-9-_.]", "");

            while(answer.contains("..")){
                answer = answer.replace("..", ".");
            }

            if(answer.startsWith(".")){
                answer = answer.replaceFirst(".", "");
            }

            if(answer.endsWith(".")){
                answer = answer.substring(0, answer.lastIndexOf("."));
            }

            if(answer.isEmpty()){
                answer = "a";
            }

            if(answer.length() > 15){
                answer = answer.substring(0, 15);

                if(answer.endsWith(".")){
                    answer = answer.substring(0, answer.lastIndexOf("."));
                }
            }

            while(answer.length() < 3) {
                answer = answer.concat(String.valueOf(answer.charAt(answer.length() - 1)));
            }

            return answer;
        }
    }
}

File: coding-test/javascript/src/kakao-blind-recruitment-2021/new-id-recommendation.js
// @ts-check

/**
 * @param {string} new_id
 * @returns {string}
 */
function solution(new_id) {
  return new_id
    .toLowerCase()
    .replace(/[^A-Za-z0-9._-]/g, '')
    .replace(/\.{2,}/g, '.')
    .replace(/^\.|\.$/g, '')
    .replace(/^$/, 'a')
    .slice(0, 15)
    .replace(/\.$/, '')
    .padEnd(3, new_id.charAt(new_id.length - 1))
}

const new_id = '...!@BaT#*..y.abcdefghijklm'
const result = 'bat.y.abcdefghi'

console.log(result === solution(new_id))
